processing/src/main/java/org/apache/druid/query/FilterDataSource.java

```diff
@@ -0,0 +1,192 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.query;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.collect.ImmutableList;
+import org.apache.druid.java.util.common.IAE;
+import org.apache.druid.query.filter.DimFilter;
+import org.apache.druid.query.planning.DataSourceAnalysis;
+import org.apache.druid.segment.FilterStorageAdapter;
+import org.apache.druid.segment.SegmentReference;
+import org.apache.druid.segment.WrappedSegmentReference;
+import org.apache.druid.utils.JvmUtils;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Function;
+
+/**
+ * This class models a data source to be unnested which is present along with a filter.
+ * An example for this data source follows:
+ *
+ * Consider this query:
+ * SELECT d3 FROM (select * from druid.numfoo where dim2='a'), UNNEST(MV_TO_ARRAY(dim3))
+ *
+ * where the filter data source has numFoo as base and dim2='a' as the filter
+ *
+ * Without this data source, the planner was converting the inner query to a query data source
+ * putting more work to be done at the broker level. This pushes the operations down to the
+ * segments and is more performant.
+ */
+public class FilterDataSource implements DataSource
+{
+
+  private final DataSource base;
+  private final DimFilter filter;
+
+  @JsonProperty("base")
+  public DataSource getBase()
+  {
+    return base;
+  }
+
+  @JsonProperty("filter")
+  public DimFilter getFilter()
+  {
+    return filter;
+  }
+
+  private FilterDataSource(DataSource base, DimFilter filter)
+  {
+    this.base = base;
+    this.filter = filter;
+  }
+
+  @JsonCreator
+  public static FilterDataSource create(
+      @JsonProperty("base") DataSource base,
+      @JsonProperty("filter") DimFilter f
+  )
+  {
+    return new FilterDataSource(base, f);
+  }
+
+  @Override
+  public Set<String> getTableNames()
+  {
+    return base.getTableNames();
+  }
+
+  @Override
+  public List<DataSource> getChildren()
+  {
+    return ImmutableList.of(base);
+  }
+
+  @Override
+  public DataSource withChildren(List<DataSource> children)
+  {
+    if (children.size() != 1) {
+      throw new IAE("Expected [1] child, got [%d]", children.size());
+    }
+
+    return new FilterDataSource(children.get(0), filter);
+  }
+
+  @Override
+  public boolean isCacheable(boolean isBroker)
+  {
+    return false;
+  }
+
+  @Override
+  public boolean isGlobal()
+  {
+    return false;
+  }
+
+  @Override
+  public boolean isConcrete()
+  {
+    return false;
```

Comment:
same question about delegating to base datasource, like iirc this returning false means it doesn't get pushed down to historicals? (i might be misremembering...)

Reply:
Yup should be base's. Done